src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php

```diff
@@ -49,82 +50,137 @@ final class AddCartRuleToOrderHandler extends AbstractOrderHandler implements Ad
     public function handle(AddCartRuleToOrderCommand $command): void
     {
         $order = $this->getOrderObject($command->getOrderId());
+        $orderInvoice = $this->createSingleOrderInvoice($order, $command);
 
-        // If the discount is for only one invoice
-        if ($order->hasInvoice() && null !== $command->getOrderInvoiceId()) {
-            $orderInvoice = new OrderInvoice($command->getOrderInvoiceId()->getValue());
-            if (!Validate::isLoadedObject($orderInvoice)) {
-                throw new OrderException('Can\'t load Order Invoice object');
+        $discountValue = (float) str_replace(',', '.', $command->getDiscountValue());
+        $cartRules = $this->buildCartRulesByType($command, $order, $discountValue, $orderInvoice);
+
+        $result = true;
+
+        foreach ($cartRules as &$cartRule) {
+            $cartRuleObj = $this->createCartRule($command, $order, $cartRule, $discountValue);
+
+            if (false === $cartRuleObj->add()) {
+                break;
```

Comment:
should we really need to break?

Reply:
what if we add nine rules, but the ten one is broken, we will have an exception, what about the nine rules?

Reply:
Fixed. (not sure why its not updating here):
But it looks like it previously was:
```
            if ($result = $cartRuleObj->add()) {
                $cartRule['id'] = $cartRuleObj->id;
            } else {
                break;
            }
```

Reply:
As I understand the command adds only one cart rule.
I think its all wrong with the naming, It should be singular.
buildCartRulesByType is only checking which type of cartRule should be built.

Ill make sure its that, and update it..

Reply:
It's disturbing `AddCartRuleToOrderHandler` should be `AddCartRulesToOrderHandler` or the `buildCartRulesByType` should be singular and receive only one cart rule (and the foreach became useless) :sweat_smile:

Reply:
yep, trying to figure out whats happening, its confusing cuz its mostly legacy code pasted to handler :smile: